tests/check-core.c

```diff
@@ -522,6 +523,66 @@ START_TEST(bootman_timeout_test)
 }
 END_TEST
 
+START_TEST(bootman_writer_simple_test)
+{
+        autofree(CbmWriter) *writer = CBM_WRITER_INIT;
+
+        fail_if(!cbm_writer_open(writer), "Failed to create writer");
+
+        cbm_writer_append(writer, "Bob");
+        cbm_writer_append(writer, "-");
+        cbm_writer_append(writer, "Jim");
+
+        fail_if(cbm_writer_error(writer) != 0, "Error should be 0");
+
+        cbm_writer_close(writer);
+        fail_if(!writer->buffer, "Failed to get writer data");
+        fail_if(!streq(writer->buffer, "Bob-Jim"), "Returned data is incorrect");
+}
+END_TEST
+
+START_TEST(bootman_writer_printf_test)
+{
+        autofree(CbmWriter) *writer = CBM_WRITER_INIT;
+
+        fail_if(!cbm_writer_open(writer), "Failed to create writer");
+
+        cbm_writer_append_printf(writer, "%s = %d", "Jim", 12);
+        fail_if(cbm_writer_error(writer) != 0, "Error should be 0");
+
+        cbm_writer_close(writer);
+        fail_if(!writer->buffer, "Failed to get writer data");
+        fail_if(!streq(writer->buffer, "Jim = 12"), "Returned data is incorrect");
+}
+END_TEST
+
+START_TEST(bootman_writer_mut_test)
+{
+        autofree(CbmWriter) *writer = CBM_WRITER_INIT;
+        char *data = NULL;
+
+        fail_if(!cbm_writer_open(writer), "Failed to create writer");
+
+        cbm_writer_append(writer, "One");
+        cbm_writer_append(writer, "Two");
+        cbm_writer_close(writer);
+        data = writer->buffer;
+        fail_if(!data, "Failed to get data");
+
+        /* Should not cause segfaults! */
+        cbm_writer_append(writer, "Three");
+        fail_if(cbm_writer_error(writer) == 0, "Error should be zero");
```

Comment:
So it makes the most sense to me to keep an "original" error. Example most of the time we'll get an error from fprintf/vfprintf. I also expect that's the code where we get OOM from (and not fclose which doesn't seem like it can return ENOMEM just ENOSPC). In the cases where (v)fprintf returns an error, it the struct's error is set to that and no other appends happen. That seems perfectly good to me.

The issue I think is the writer->memstream check in the appends. This could happen either because open wasn't called or if open_memstream failed. In the latter case I think we should set ENOMEM to error but otherwise not do anything. Then we can say for sure if we get to the check in an append and memstream is NULL then the writer wasn't opened and can say EBADF.

Thoughts?